src/types/Diagnostic.ts

```diff
@@ -1,67 +1,12 @@
 import { Lintable } from "./Lintable";
-import {
-  SyncRuleId,
-  AsyncRuleId,
-  RuleId,
-  AsyncRuleConfig,
-  SyncRuleConfig,
-} from "./RuleConfig";
+import { RuleId, RuleConfig } from "./RuleConfig";
 
-type BaseDiagnostic = {
+export type Diagnostic<
+  R extends RuleId = RuleId,
+  C extends RuleConfig = RuleConfig
+> = {
   id: string;
   lintable: Lintable;
   fix?(): void;
-};
-
-/**
- * Essentially the old Diagnostic type
- */
-export type SyncDiagnostic<
-  R extends SyncRuleId = SyncRuleId,
-  C extends SyncRuleConfig = SyncRuleConfig extends infer A
-    ? A extends { rule: R }
-      ? SyncRuleConfig
-      : never
-    : never
-> = BaseDiagnostic & {
-  type: "sync";
-  rule: R;
-} & C["info"];
-
-/**
- * New Diagnostic type that can be Pending
- * Although `hasViolation` can be false, Design Linter should not expose
- * those Diagnostics to consumers
- */
-export type AsyncDiagnostic<
-  R extends AsyncRuleId = AsyncRuleId,
-  C extends AsyncRuleConfig = AsyncRuleConfig extends infer A
-    ? A extends { rule: R }
-      ? AsyncRuleConfig
-      : never
-    : never
-> = BaseDiagnostic & {
-  type: "async";
   rule: R;
-  hasViolation: boolean | "pending";
 } & C["info"];
```

Comment:
Since configs are optional

```suggestion
  config?: C;
};
```